### PR TITLE
feat: allow user to provide TS program instance in parser options

### PR DESCRIPTION
**Pull request by uniqueiniquity**
`typescript-estree` has complex handling for constructing and managing TS program instances in order to provide type info for semantic rules. However, for one-off invocations on large codebases (e.g., in CI), this functionality isn't valuable, and the overhead is substantial.

Here, we introduce a new parser option `program` that allows users to programmatically provide their own TS program instance that will be used in rules for type info. This is based on [TSLint's functionality](https://github.com/palantir/tslint/blob/285fc1db18d1fd24680d6a2282c6445abf1566ee/src/linter.ts#L121).

In this initial implementation, we only allow for a single program to be provided. However, this could be augmented in the future to support multiple programs.

Fixes https://github.com/typescript-eslint/typescript-eslint/issues/1442

### CODE REVIEW

File: packages/typescript-estree/package.json [hunk 1]
```diff
@@ -64,6 +64,9 @@
     "jest-specific-snapshot": "*",
     "make-dir": "*",
     "tmp": "^0.2.1",
+    "typescript": "^4.3.2"
+  },
+  "peerDependencies": {
```
Comment (bradzacher) [inline packages/typescript-estree/package.json:69]: We have an implicit peer dependency via the `peerDependenciesMeta` property.

The reason we don't explicitly declare the peer dependency is because on older versions of yarn and npm, they do not understand `peerDependenciesMeta`.
So if we add an explicit peer dep - then they will log warnings at install time.

In the past this annoyed many users 😅 that had our packages installed (but not used) transitively.

TL;DR - remove the explicit peer dep

File: packages/types/package.json [hunk 2]
```diff
@@ -48,5 +48,8 @@
         "_ts3.4/*"
       ]
     }
+  },
+  "devDependencies": {
+    "typescript": ">=2.6.0"
```
Comment (bradzacher) [inline packages/types/package.json:53]: We use the root package.json to control the TS version installed in the repo.

You can use the `peerDependenciesMeta` to add an optional peer dep on typescript, which we'll need to do to ensure package managers can understand the new requirement

```suggestion
    "typescript": "*"
```

File: packages/types/src/parser-options.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import { Lib } from './lib';
+import { Program } from 'typescript';
```
Comment (bradzacher) [inline packages/types/src/parser-options.ts:2]: lets ensure that this isn't ever pulled in as a runtime dep on this package.

```suggestion
import type { Program } from 'typescript';
```

File: packages/typescript-estree/src/create-program/useProvidedProgram.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  extra: Extra,
+): ASTAndProgram | undefined {
+  log('Retrieving ast for %s from provided program instance', extra.filePath);
+
+  const astAndProgram = getAstFromProgram(programInstance, extra);
+
+  if (!astAndProgram) {
+    const errorLines = [
+      '"parserOptions.program" has been provided for @typescript-eslint/parser.',
+      `The file was not found in the provided program instance: ${extra.filePath}`,
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/useProvidedProgram.ts:19]: The filepath should be made relative.
This ensures that error messages are kept is stable and portable.

https://github.com/typescript-eslint/typescript-eslint/blob/fc5f171b1ade2b45a1a4268b6d22926d420282a6/packages/typescript-estree/src/create-program/createProjectProgram.ts#L52-L55

File: packages/typescript-estree/src/parser.ts [hunk 5]
```diff
@@ -264,22 +269,32 @@ function applyParserOptionsToExtra(options: TSESTreeOptions): void {
...
-    projectFolderIgnoreList,
-  );
+  if (typeof options.program === 'object') {
+    extra.program = options.program;
+    log(
+      'parserOptions.program was provided, so parserOptions.project will be ignored.',
+    );
+  }
+
+  if (extra.program === null) {
```
Comment (bradzacher) [inline packages/typescript-estree/src/parser.ts:279]: nit - just for safety so we can handle the undefined case as well, in case something somewhere is wacky.

```suggestion
  if (extra.program == null) {
```

File: packages/typescript-estree/src/parser.ts [hunk 6]
```diff
@@ -264,22 +269,32 @@ function applyParserOptionsToExtra(options: TSESTreeOptions): void {
...
-      return acc;
-    }, [])
-    // prefix with a ! for not match glob
-    .map(folder => (folder.startsWith('!') ? folder : `!${folder}`));
-  // NOTE - prepareAndTransformProjects relies upon having the correct tsconfigRootDir in extra
-  extra.projects = prepareAndTransformProjects(
-    options.project,
-    projectFolderIgnoreList,
-  );
+  if (typeof options.program === 'object') {
```
Comment (bradzacher) [inline packages/typescript-estree/src/parser.ts:272]: nit - we don't want to log if the user explicitly passes `null` here.

```suggestion
  if (typeof options.program === 'object' && options.program != null) {
```

**Review by bradzacher — CHANGES_REQUESTED** [review]
i didn't review the tests - just glanced through the code.
looks good! a few comments.

have you tested this in a real-world lint setup?
I'd be interested to know how much this improves the perf for them.
Also would like to triple check that ESLint doesn't mangle the program instance or anything.

**Comment by uniqueiniquity** [comment]
@bradzacher I'm about to try it on a large project - will report back alongside addressing your suggestions.

**Comment by uniqueiniquity** [comment]
Closing in favor of #3484 - I screwed up my branch casing and couldn't fix it 😫
I've addressed all comments currently on this PR and added perf results on the new PR - spoiler alert, they're good.

File: packages/typescript-estree/src/parser.ts (hunk 6 above)
Comment (uniqueiniquity) [reply packages/typescript-estree/src/parser.ts:272]: I made this `if (options.program && typeof options.program === 'object')`

File: packages/typescript-estree/src/parser.ts (hunk 5 above)
Comment (uniqueiniquity) [reply packages/typescript-estree/src/parser.ts:279]: I made this `if (!extra.program)`
